### PR TITLE
Several gems need at least some value for total_allocated_objects

### LINKED ISSUE #1681: GC.stat :total_allocated_objects is not supported
I'm not sure if TruffleRuby is able to support this or Rails needs to change this in code here

https://github.com/rails/rails/blob/6-0-stable/activesupport/lib/active_support/notifications/instrumenter.rb#L159

But there is a specific case for JRuby where this part returns `0` instead.

This causes the following error in across Rails 6 tests in `ActiveSupport::Notification.instrument`

```
ArgumentError: wrong number of arguments (given 1, expected 0)
    ~/dev/bugs/truffleruby/rails/activesupport/lib/active_support/notifications/instrumenter.rb:159:in `stat'
    ~/dev/bugs/truffleruby/rails/activesupport/lib/active_support/notifications/instrumenter.rb:159:in `now_allocations'
    ~/dev/bugs/truffleruby/rails/activesupport/lib/active_support/notifications/instrumenter.rb:81:in `start!'
    ~/dev/bugs/truffleruby/rails/activesupport/lib/active_support/subscriber.rb:132:in `start'
    ~/dev/bugs/truffleruby/rails/activesupport/lib/active_support/log_subscriber.rb:103:in `start'
    ~/dev/bugs/truffleruby/rails/activesupport/lib/active_support/notifications/fanout.rb:156:in `start'
    ~/dev/bugs/truffleruby/rails/activesupport/lib/active_support/notifications/fanout.rb:58:in `block in start'
    ~/dev/bugs/truffleruby/rails/activesupport/lib/active_support/notifications/fanout.rb:58:in `each'
    ~/dev/bugs/truffleruby/rails/activesupport/lib/active_support/notifications/fanout.rb:58:in `start'
    ~/dev/bugs/truffleruby/rails/activesupport/lib/active_support/notifications/instrumenter.rb:36:in `start'
    ~/dev/bugs/truffleruby/rails/activesupport/lib/active_support/notifications/instrumenter.rb:22:in `instrument'
    ~/dev/bugs/truffleruby/rails/activesupport/lib/active_support/notifications.rb:180:in `instrument'
    ~/dev/bugs/truffleruby/rails/actionpack/lib/action_controller/metal/instrumentation.rb:30:in `process_action'
    ~/dev/bugs/truffleruby/rails/actionpack/lib/action_controller/metal/params_wrapper.rb:245:in `process_action'
    ~/dev/bugs/truffleruby/rails/actionpack/lib/abstract_controller/base.rb:136:in `process'
    ~/dev/bugs/truffleruby/rails/actionview/lib/action_view/rendering.rb:39:in `process'
    ~/dev/bugs/truffleruby/rails/actionpack/lib/action_controller/metal.rb:191:in `dispatch'
    ~/dev/bugs/truffleruby/rails/actionpack/lib/action_controller/test_case.rb:513:in `process'
    ~/dev/bugs/truffleruby/rails/actionpack/lib/action_controller/test_case.rb:400:in `post'
    ~/dev/bugs/truffleruby/rails/actionpack/test/controller/request_forgery_protection_test.rb:778:in `block in test_should_not_warn_if_form_authenticity_param_matches_form_authenticity_token'
```

### PR DESCRIPTION
**Pull request by chrisseaton**
Fixes #1681.

For example `activesupport` and `shopify_metrics` (internal gem). `activesupport` has logic to detect JRuby and give it the value `0`, so we might as well just report `0` in the first place.

### CODE REVIEW

File: spec/ruby/core/gc/stat_spec.rb [hunk 1]
```diff
@@ -1,19 +1,16 @@
 require_relative '../../spec_helper'
 
 describe "GC.stat" do
-  it "supports access by key" do
-    keys = [:heap_free_slots, :total_allocated_objects, :count]
-    keys.each do |key|
-      GC.stat(key).should be_kind_of(Integer)
-    end
-  end
```
Comment (eregon) [inline spec/ruby/core/gc/stat_spec.rb:9]: Should we keep this spec? It should pass now.
Notably it tests that all 3 keys are there.

File: src/main/ruby/truffleruby/core/gc.rb [hunk 2]
```diff
@@ -76,31 +76,32 @@ def garbage_collect
...
       peak_max: 0,
-      last_used: 0,
-      last_committed: 0,
-      last_init: 0,
-      last_max: 0,
+      peak_used: 0,
+      time: time,
+      total_allocated_objects: 0,
+      unknown_count: unknown_count, # if nonzero, major or minor count needs to be updated for this GC case
+      used: used,
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/gc.rb:104]: Could you undo the reordering? The order is intentionally close to what MRI has and also makes it easy by having `*used` and `*committed` next to each other.

File: spec/ruby/core/gc/stat_spec.rb (hunk 1 above)
Comment (chrisseaton) [reply spec/ruby/core/gc/stat_spec.rb:9]: I added three separate specs for the three keys we were tracking. I think separate specs is better, as you can then improve each other time (fore example some might have sensible values.)

**Comment by chrisseaton** [comment]
Yes `GC.stat[:total_allocated_objects]` being an integer. But `GC.stat(:foo)` returning `0` is wrong we should fix that!

File: spec/ruby/core/gc/stat_spec.rb (hunk 1 above)
Comment (eregon) [reply spec/ruby/core/gc/stat_spec.rb:9]: I see, that's fine now.

**Comment by eregon** [comment]
> Yes `GC.stat[:total_allocated_objects]` being an integer.

OK, could you add specs for that as well then?

> But `GC.stat(:foo)` returning `0` is wrong we should fix that!

I'm not sure, I think it's kind of fine, does it cause any issue?

File: spec/ruby/core/gc/stat_spec.rb [hunk 3]
```diff
@@ -1,19 +1,16 @@
...
-
   it "returns hash of values" do
     stat = GC.stat
     stat.should be_kind_of(Hash)
     stat.keys.should include(:count)
   end
 
+  it "can return a single value" do
+    GC.stat(:count).should be_kind_of(Integer)
+  end
```
Comment (eregon) [inline spec/ruby/core/gc/stat_spec.rb:12]: This duplicates the `provides some number for count` spec below.

**Comment by chrisseaton** [comment]
> I'm not sure, I think it's kind of fine, does it cause any issue?

Well for one thing it's making `GC.stat(x)` behave differently to `GC.stat[x]`, which you're now asking to me to spec separately since they behave differently 🙃

**Comment by chrisseaton** [comment]
I've pushed an extra commit on top with how I think it should be.

Possibly also want to short-cut `GC.stat(x)` for stats available without doing the pool-specific stats later on, but not now.

File: spec/ruby/core/gc/stat_spec.rb (hunk 3 above)
Comment (chrisseaton) [reply spec/ruby/core/gc/stat_spec.rb:12]: I don't mind that - I focus on what the specification specifies and I think it's ok if it specifies two different things using the same code to do it.

**Review by eregon — APPROVED** [review]

**Pull request merged**